### PR TITLE
Implemented grindstones

### LINKED ISSUE #501: Implement grindstones

### PR DESCRIPTION
**Pull request by JustTalDevelops**
Resolves #501.

### CODE REVIEW

**Review by DaPigGuy — APPROVED** [review]

File: server/block/grindstone.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+// Grindstone is a block that repairs items and tools as well as removing enchantments from them. It also serves as a
+// weaponsmith's job site block.
+type Grindstone struct {
+	transparent
+
+	// Attach represents the attachment type of the Grindstone.
+	Attach GrindstoneAttachment
+	// Direction represents the direction the Grindstone is facing.
+	Direction cube.Direction
```
Comment (Sandertv) [inline server/block/grindstone.go:20]: Would call this field `Facing` for consistency.

File: server/block/grindstone.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+// UseOnBlock ...
+func (g Grindstone) UseOnBlock(pos cube.Pos, face cube.Face, _ mgl64.Vec3, w *world.World, user item.User, ctx *item.UseContext) (used bool) {
+	pos, face, used = firstReplaceable(w, pos, face, g)
+	if !used {
+		return false
+	}
+	if face == cube.FaceUp {
+		g.Direction = user.Facing().Opposite()
```
Comment (Sandertv) [inline server/block/grindstone.go:44]: Can move `g.Direction = user.Facing().Opposite()` in front of the `if`s and only change it in the `else` case to save a line.

File: server/item/stack.go [hunk 3]
```diff
@@ -248,6 +248,9 @@ func (s Stack) WithoutEnchantments(enchants ...EnchantmentType) Stack {
 	for _, enchant := range enchants {
 		delete(s.enchantments, reflect.TypeOf(enchant))
 	}
+	if _, ok := s.item.(EnchantedBook); ok && len(s.enchantments) == 0 {
+		s.item = Book{}
+	}
```
Comment (Sandertv) [inline server/item/stack.go:253]: Out of scope but I'll accept it this time :wink:

File: server/session/handler_anvil.go [hunk 4]
```diff
@@ -180,9 +180,9 @@ func (h *ItemStackRequestHandler) handleCraftRecipeOptional(a *protocol.CraftRec
 	return nil
 }
 
-// repairWithMaterial is a helper function that repairs an item stack with a given material stack. It returns the new item
+// anvilRepairWithMaterial is a helper function that repairs an item stack with a given material stack. It returns the new item
 // stack, the cost, and the repaired items count.
-func repairWithMaterial(input item.Stack, material item.Stack, result item.Stack) (item.Stack, int, int, error) {
+func anvilRepairWithMaterial(input item.Stack, material item.Stack, result item.Stack) (item.Stack, int, int, error) {
```
Comment (Sandertv) [inline server/session/handler_anvil.go:185]: I think the previous name was better, maybe `repairItemWithMaterial`.

File: server/session/handler_anvil.go [hunk 5]
```diff
@@ -199,8 +199,8 @@ func repairWithMaterial(input item.Stack, material item.Stack, result item.Stack
 	return result, cost, count, nil
 }
 
-// repairWithDurable is a helper function that repairs an item with another durable item stack.
-func repairWithDurable(input item.Stack, durable item.Stack, result item.Stack) (item.Stack, int) {
+// anvilRepairWithDurable is a helper function that repairs an item with another durable item stack.
+func anvilRepairWithDurable(input item.Stack, durable item.Stack, result item.Stack) (item.Stack, int) {
```
Comment (Sandertv) [inline server/session/handler_anvil.go:203]: I think the previous name was better, maybe `repairItemWithDurable`.

File: server/session/handler_anvil.go [hunk 6]
```diff
@@ -215,9 +215,9 @@ func repairWithDurable(input item.Stack, durable item.Stack, result item.Stack)
 	return result, cost
 }
 
-// mergeEnchantments merges the enchantments of the material item stack onto the result item stack and returns the result
+// anvilMergeEnchantments merges the enchantments of the material item stack onto the result item stack and returns the result
 // item stack, booleans indicating whether the enchantments had any compatible or incompatible enchantments, and the cost.
-func mergeEnchantments(input item.Stack, material item.Stack, result item.Stack, cost int, enchantedBook bool) (item.Stack, bool, bool, int) {
+func anvilMergeEnchantments(input item.Stack, material item.Stack, result item.Stack, cost int, enchantedBook bool) (item.Stack, bool, bool, int) {
```
Comment (Sandertv) [inline server/session/handler_anvil.go:220]: Again, I prefer the old name. Any particular reason to prefix this all with `anvil`?

File: server/session/handler_grindstone.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+	if firstInput.Empty() && secondInput.Empty() {
+		return fmt.Errorf("input item(s) are empty")
+	}
+	if firstInput.Count() > 1 || secondInput.Count() > 1 {
+		return fmt.Errorf("input item(s) are not single items")
+	}
+
+	resultStack := existingItem(firstInput, secondInput)
+	if !firstInput.Empty() && !secondInput.Empty() {
+		resultStack = firstInput.WithEnchantments(secondInput.Enchantments()...)
```
Comment (Sandertv) [inline server/session/handler_grindstone.go:49]: Is this correct? Isn't the output item always supposed to be enchantment free?

File: server/session/handler_grindstone.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+	for _, enchant := range stack.Enchantments() {
+		// TODO: Don't remove curses.
+		stack = stack.WithoutEnchantments(enchant.Type())
+	}
+	return stack
+}
+
+// existingItem returns the item.Stack that exists out of two input items. The function expects at least one of the
+// items to be non-empty.
+func existingItem(first, second item.Stack) item.Stack {
```
Comment (Sandertv) [inline server/session/handler_grindstone.go:106]: I didn't really get the meaning of this without the documentation. Maybe something like `nonZeroItem()` would be better?

File: server/session/handler_grindstone.go (hunk 7 above)
Comment (Sandertv) [reply server/session/handler_grindstone.go:49]: Ah I see, this is to get the total experience from all those enchantments later on. Might want to document that.

**Review by Sandertv — CHANGES_REQUESTED** [review]
Got a couple of comments. Mostly some small code/renaming improvements or documentation. Should be ready to go once fixed.

File: server/session/handler_anvil.go (hunk 6 above)
Comment (JustTalDevelops) [reply server/session/handler_anvil.go:220]: No reason - was a change that I made while testing and forgot to revert.

**Comment by JustTalDevelops** [comment]
Resolved all comments.

**Review by Sandertv — APPROVED** [review]
This looks good to me. Thanks for the PR!

**Pull request merged**